### PR TITLE
[10.x] Fix infinite loop when global scopes query contains aggregates

### LINKED ISSUE #49968: `Error: Xdebug has detected a possible infinite loop, and aborted your script with a stack depth of '512' frames`
### Laravel Version

10.43.0

### PHP Version

8.1.27

### Database Driver & Version

MySQL version: 8.0.35

### Description

1. Our internal CI/CD started failing out of the blue with `Error: Xdebug has detected a possible infinite loop, and aborted your script with a stack depth of '512' frames`

2. After some work, we managed to reproduce at https://github.com/mxts/laravel-issue, if using `10.43.0`, the test will fail, while test will pass if using `10.42.0`

3. The issue comes from the internals and we still have not narrowed down the code, but when our entire scope here https://github.com/mxts/laravel-issue/blob/b6d93fd817c6e0ed9d88c20e157137a1a7af09f9/app/Models/WithTx.php#L9-L15 is commented out, the problem goes away

4. Not quite sure why our scope leads Laravel to nest infinitely

### Steps To Reproduce

1. Clone https://github.com/mxts/laravel-issue
2. Modify `.env.testing` with DB credentials
3. Run `composer update` to get `10.43.0` and watch `php artisan test` fail
4. Modify `composer.json` to use `10.42.0` 
5. Run `composer update` to get `10.42.0` and watch `php artisan test` pass

### PR DESCRIPTION
**Pull request by mateusjunges**
Attempt to fix #49968.

>TLDR: We can't call `$this->getGrammar()` directly because this causes an infinite loop.

Pinging @tpetry here because this is related to #49912. I'm not sure how to test this TBH, so I'd appreciate your input here. 

Calling `$this->getGrammar()` will redirect the call (passthru) from `\Illuminate\Database\Eloquent\Builder` to the base `Illuminate\Database\Query\Builder`. When doing this, we add a call to `->applyScopes()`:

https://github.com/laravel/framework/blob/4f7802dfc9993cb57cf69615491ce1a7eb2e9529/src/Illuminate/Database/Eloquent/Builder.php#L1760

This will cause laravel to call the scopes we have defined, and, since we are using a `count` aggregate, we'll end up in the  `withAggregate` method defined on line  606 of `Illuminate\Database\Eloquent\Concerns\QueriesRelationships` trait. Now, since we are calling `$this->getGrammar()` directly, we'll start the process of forwarding the call to the base query builder and calling `->applyScopes` all over again, which will cause an infinite loop.

I hope I was clear enough, but help to explain in the comments if need be.

### CODE REVIEW

**Comment by mateusjunges** [comment]
I created a repo to reproduce the issue: https://github.com/mateusjunges/reproduce-laravel-49968. Just install the repo locally and run the tests.

**Comment by bilogic** [comment]
@mateusjunges This is a life saver!

**Comment by bilogic** [comment]
@mateusjunges This is a life saver!

**Comment by tpetry** [comment]
Sorry @mateusjunges. I am always embarrassed when breaking something. I didn‘t think on this edge case when writing that patch. There are so many ways you can break the database part and many code paths don‘t have a unit test. Can you PR a unit test for this case so we have some guarantee we can‘t break this behaviour in the future again with some other changes.

**Comment by mxts** [comment]
> This will cause laravel to call the scopes we have defined, and, since we are using a `count` aggregate, we'll end up in the `withAggregate` method defined on line 606 of `Illuminate\Database\Eloquent\Concerns\QueriesRelationships` trait. Now, since we are calling `$this->getGrammar()` directly, we'll start the process of forwarding the call to the base query builder and calling `->applyScopes` all over again, which will cause an infinite loop.

wow thanks @mateusjunges 

Did you use any particular method to track this down? I'm just so amazed at how quickly you fixed it when it took me almost 3 days just to replicate it.

**Comment by mateusjunges** [comment]
@mxts it's basically tracking it down to which PR broke the expected behavior. I usually do it installing the project locally and the you can use Git to go back in time to previous versions of the code. Once you know which PR caused the issue it gets easier as  you have the diff. This gives me a  general idea about what could be the issue. Then I use xdebug to track it down to the root cause.
